### PR TITLE
[RDY] Welcome screen

### PR DESCRIPTION
**Pull request by ohuang12**
Adds collective directory for all databases.
Scans this directory for other directories and lists them in the welcome screen.
Double clicking/loading an item in the list, will cause to load the DB: pressing start creates the split-pane.
Opening file creates own DB in the collective directory.
Selecting new DB path, scans this directory for other directories and lists them.
Edit: File opening now requires loading in both graph and newick file.

Notes:
~~- File opening now depends on having a phylogenetic tree loading in.~~

### CODE REVIEW

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/FileOpenController.java [hunk 1]
```diff
@@ -58,6 +62,8 @@
 	private ProgressDialog progressDialog;
 	private ObjectProperty<TreeNode> treeProperty;
 	private ObjectProperty<Graph> graphProperty;
+	private ObjectProperty<ObservableList<String>> dbPathProperty;
+	private BooleanProperty done = new SimpleBooleanProperty(false, "done");;
```
Comment (Unia) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/FileOpenController.java]: Double `;`

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WelcomeController.java [hunk 2]
```diff
@@ -1,37 +1,174 @@
...
 public class WelcomeController {
-	private BooleanProperty done = new SimpleBooleanProperty(false, "done");
-	@SuppressWarnings("unused") @FXML
-	private ListView list;
-	@SuppressWarnings("unused") @FXML
-	private void startButtonAction(ActionEvent e) {
+	private static final String SELECT_OPTION = "Select new database...";
+	private ObservableList<String> items;
+	private ObjectProperty<Graph> dbProperty;
+	private BooleanProperty done = new SimpleBooleanProperty(false, "done");;
```
Comment (Unia) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WelcomeController.java]: Double `;`

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WelcomeController.java [hunk 3]
```diff
@@ -1,37 +1,174 @@
...
+		} catch (Exception e) {
+			e.printStackTrace();
+		}
+	}
+	
+	@SuppressWarnings("unused") @FXML
+	private void loadButtonAction(ActionEvent e) {
+		progressDialog = new ProgressDialog(list.getParent());
+		if (dbpath == SELECT_OPTION) {
+			File dir = selectDirectory();
```
Comment (Unia) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WelcomeController.java]: If `dir` is not to be used anywhere, you might as well do `if (selectDirectory() == null)`

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WelcomeController.java [hunk 4]
```diff
@@ -1,37 +1,174 @@
...
+	}
 
 	@SuppressWarnings("unused") @FXML
-	private void initialize() {
-		ObservableList<String> items = FXCollections.observableArrayList(
-				"nl/tudelft/DNAinator/db/10_strains",
-				"nl/tudelft/DNAinator/db/38_strains",
-				"nl/tudelft/DNAinator/db/100_strains");
+	private void onMouseClicked(MouseEvent e) {
+		if (e.getClickCount() == 2) {
```
Comment (Unia) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WelcomeController.java]: Are we sure we want a double click, when everything else in the UI works with a single click?

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WelcomeController.java [hunk 5]
```diff
@@ -1,37 +1,174 @@
...
+			if (newV != SELECT_OPTION) {
+				deleteButton.setDisable(false);
+			}
+			dbload.setDatabase(newV);
+			setDBPath();
+			System.out.println(this.dbpath);
+		});
+	}
+	
+	private void setDBPath(){
```
Comment (Unia) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WelcomeController.java]: Why is this needed?

Comment (ohuang12) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WelcomeController.java]: This is not necessary, remains from old progress.

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WelcomeController.java (hunk 4 above)
Comment (ohuang12) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WelcomeController.java]: I thought this would be intuitive for opening/loading something from a list.

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WelcomeController.java (hunk 3 above)
Comment (ohuang12) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WelcomeController.java]: Yes agreed, will update.

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/FileOpenController.java (hunk 1 above)
Comment (ohuang12) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/FileOpenController.java]: Whoops! Will remove.

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WelcomeController.java (hunk 4 above)
Comment (pvanagtmaal) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WelcomeController.java]: It is, would be stupid if you accidentally load the 671 strains when wanted to select 100

Comment (Unia) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WelcomeController.java]: Oke :)

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WelcomeController.java [hunk 6]
```diff
@@ -1,37 +1,165 @@
...
+	}
+	
+	/**
+	 * Scan the directory containing the default location of databases.
+	 * Adds all the directories found to the welcomescreen's list of selectables.
+	 * @throws IOException
+	 */
+	private void scanDirectory() throws IOException {
+		try (DirectoryStream<Path> ds = Files.newDirectoryStream(Paths.get(DEFAULT_DB_PATH))) {
+	        for (Path path : ds) {
```
Comment (pvanagtmaal) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WelcomeController.java]: Bad indentation

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WindowController.java [hunk 7]
```diff
@@ -31,6 +31,9 @@
 	@SuppressWarnings("unused") @FXML
 	private void initialize() {
 		ColorServer colorServer = new ColorServer();
+		fileOpenController.dbPathProperty().set(welcomeController.getListedPaths());
+		fileOpenController.doneProperty().addListener((obj, oldV, newV) -> 
+			welcomeController.doneProperty().setValue(true));
```
Comment (pvanagtmaal) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WindowController.java]: `newV` is always `true`?

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/FileOpenController.java [hunk 8]
```diff
@@ -58,6 +60,8 @@
 	private ProgressDialog progressDialog;
 	private ObjectProperty<TreeNode> treeProperty;
 	private ObjectProperty<Graph> graphProperty;
+	private ObjectProperty<ObservableList<String>> dbPathProperty;
```
Comment (pvanagtmaal) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/FileOpenController.java]: A `ListProperty` does not suffice?

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WelcomeController.java [hunk 9]
```diff
@@ -1,37 +1,165 @@
...
+import javafx.scene.control.Button;
 import javafx.scene.control.ListView;
+import javafx.scene.input.MouseEvent;
+import javafx.stage.DirectoryChooser;
 
 /**
  * Controller for the welcome screen.
  */
 public class WelcomeController {
+	private static final String SELECT_OPTION = "Select new database...";
```
Comment (pvanagtmaal) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WelcomeController.java]: Can this be set in the FXML?

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WelcomeController.java [hunk 10]
```diff
@@ -1,37 +1,165 @@
...
 		done.setValue(true);
 	}
+	
+	@SuppressWarnings("unused") @FXML 
+	private void deleteButtonAction(ActionEvent a) {
+		try {
+			FileUtils.deleteRecursively(new File(getDBPath()));
+			items.remove(getDBPath());
+		} catch (Exception e) {
+			e.printStackTrace();
```
Comment (pvanagtmaal) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WelcomeController.java]: Should the user be notified that the database could not be deleted?

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WelcomeController.java [hunk 11]
```diff
@@ -1,37 +1,165 @@
...
+				list.getSelectionModel().select(getDBPath());
+			} else {
+				progressDialog = new ProgressDialog(list.getParent());
+				dbload.restart();
+			}
+		}
+	}
+	
+	@SuppressWarnings("unused") @FXML
+	private void initialize() throws IOException {
```
Comment (pvanagtmaal) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WelcomeController.java]: Where is this exception caught?

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WelcomeController.java [hunk 12]
```diff
@@ -1,37 +1,165 @@
...
+			deleteButton.setDisable(true);
+			if (newV != SELECT_OPTION) {
+				deleteButton.setDisable(false);
+			}
+			dbload.setDatabase(newV);
+			System.out.println(this.getDBPath());
+		});
+	}
+	
+	private String getDBPath() {
```
Comment (pvanagtmaal) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WelcomeController.java]: private getter?

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WelcomeController.java (hunk 11 above)
Comment (ohuang12) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WelcomeController.java]: Still has to be caught somewhere, I'll update to catch it in the method where the IO happens.

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WelcomeController.java (hunk 10 above)
Comment (ohuang12) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WelcomeController.java]: Yes, will update.

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WindowController.java (hunk 7 above)
Comment (ohuang12) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WindowController.java]: Yes newV will be always true.

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WelcomeController.java (hunk 6 above)
Comment (ohuang12) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WelcomeController.java]: Will update.

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WelcomeController.java (hunk 9 above)
Comment (ohuang12) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WelcomeController.java]: Looks like it can. Will move it to the FXML.

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WelcomeController.java (hunk 12 above)
Comment (ohuang12) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WelcomeController.java]: Getter is only used inside the welcome screen and will not be used outside it (method is only used to make code look cleaner).

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/FileOpenController.java (hunk 8 above)
Comment (ohuang12) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/FileOpenController.java]: Yes it will, changing.

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WelcomeController.java [hunk 13]
```diff
@@ -1,37 +1,167 @@
...
 	private BooleanProperty done = new SimpleBooleanProperty(false, "done");
-	@SuppressWarnings("unused") @FXML
-	private ListView list;
-	@SuppressWarnings("unused") @FXML
-	private void startButtonAction(ActionEvent e) {
+	private DBLoadService dbload;
+	private DirectoryChooser dirChooser;
+	private ProgressDialog progressDialog;
+	private ObservableList<String> items;
+	private static final String DEFAULT_DB_PATH = "target\\db";
```
Comment (Balletie) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WelcomeController.java]: This is only compatible with Windows

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/services/GraphLoadService.java [hunk 14]
```diff
@@ -25,8 +26,7 @@
  * </p>
  */
 public class GraphLoadService extends Service<Graph> {
-	private static final String DB_PATH = "target/dna-graph-db";
-	
+	private static final String DB_PATH = "target\\db\\db";
```
Comment (Balletie) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/services/GraphLoadService.java]: As above, please use `FileSystem.getSeparator`

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/services/DBLoadService.java [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+import javafx.concurrent.Service;
+import javafx.concurrent.Task;
+import nl.tudelft.dnainator.graph.Graph;
+import nl.tudelft.dnainator.graph.impl.Neo4jGraph;
+
+/**
+ * A JavaFX service that loads an existing neo4j database based on its directory path.
+ */
+public class DBLoadService extends Service<Graph> {
+	private static final String DEFAULT_DB_PATH = "target\\db";
```
Comment (Balletie) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/services/DBLoadService.java]: `FileSystem.getSeparator`, same as other comments.

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/services/DBLoadService.java [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+	private ObjectProperty<String> database = new SimpleObjectProperty<>(this, "database");
+
+	/**
+	 * Construct a DBLoadService with a default database path.
+	 */
+	public DBLoadService() {
+		setDatabase(DEFAULT_DB_PATH);
```
Comment (Balletie) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/services/DBLoadService.java]: Consider passing the path via the constructor.

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/FileOpenController.java [hunk 17]
```diff
@@ -76,6 +91,7 @@ private void initialize() {
 		graphLoadService.setOnRunning(e -> progressDialog.show());
 		graphLoadService.setOnSucceeded(e -> {
 			graphProperty.setValue(graphLoadService.getValue());
+			done.setValue(!done.getValue());
```
Comment (gfokkema) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/FileOpenController.java]: shouldn't this be done.setValue(true)?

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/FileOpenController.java [hunk 18]
```diff
@@ -91,21 +107,17 @@ private void initialize() {
...
 	private void bindDisabledFieldsAndButtons() {
 		BooleanBinding emptyGraphFile = nodeField.textProperty().isEmpty()
-				.or(edgeField.textProperty().isEmpty());
+				.or(edgeField.textProperty().isEmpty()).or(newickField.textProperty().isEmpty());
 		gffField.disableProperty().bind(graphProperty.isNull().and(emptyGraphFile));
-		// At least both graph files or the newick file must be filled.
-		openButton.disableProperty().bind(emptyGraphFile
-				.and(newickField.textProperty().isEmpty())
-				.and(gffField.disableProperty()));
+		// At least both graph files and the newick file must be filled.
```
Comment (gfokkema) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/FileOpenController.java]: Because of the change in annotation batchinsert, gff files need to be loaded as well

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/FileOpenController.java (hunk 17 above)
Comment (ohuang12) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/FileOpenController.java]: This is maybe done a little dirty, but right now I flip the property to enable loading new files after an existing one is loaded (since the property is notified, it will create a new split pane again).

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/FileOpenController.java (hunk 18 above)
Comment (ohuang12) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/FileOpenController.java]: Yes, will change it.

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/FileOpenController.java [hunk 19]
```diff
@@ -23,13 +27,12 @@
 import nl.tudelft.dnainator.javafx.widgets.animations.TransitionAnimation.Position;
 import nl.tudelft.dnainator.javafx.widgets.dialogs.ExceptionDialog;
 import nl.tudelft.dnainator.javafx.widgets.dialogs.ProgressDialog;
-import org.neo4j.io.fs.FileUtils;
```
Comment (Unia) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/FileOpenController.java]: The frontend now does not depend on Neo4j anymore. As such, this dependency should be removed from the parent pom and placed in `dnainator-javafx/pom.xml` instead.

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WelcomeController.java [hunk 20]
```diff
@@ -1,37 +1,165 @@
...
+	private static final String DEFAULT_DB_PATH = "target" + File.separator + "db";
+	@SuppressWarnings("unused") @FXML private Button deleteButton;
+	@SuppressWarnings("unused") @FXML private Button loadButton;
+	@SuppressWarnings("unused") @FXML private ListView<String> list;
+	@SuppressWarnings("unused") @FXML private String selectDB;
+	
+	@SuppressWarnings("unused") @FXML 
+	private void deleteButtonAction(ActionEvent a) {
+		try {
+			FileUtils.deleteRecursively(new File(getDBPath()));
```
Comment (Unia) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WelcomeController.java]: Disregard my above comment on the neo4j dependency.

**Comment by Unia** [comment]
Looking good now! :) Second new reviewer may merge.

File: dnainator-javafx/src/main/resources/fxml/welcome.fxml [hunk 21]
```diff
@@ -1,20 +1,30 @@
...
+      </ListView>
+      <Button fx:id="loadButton" layoutX="457.0" layoutY="343.0" onAction="#loadButtonAction" text="Load" AnchorPane.bottomAnchor="32.0" AnchorPane.rightAnchor="101.0" />
+        <Button fx:id="deleteButton" disable="true" layoutX="509.0" layoutY="343.0" onAction="#deleteButtonAction" text="Delete" AnchorPane.bottomAnchor="32.0" AnchorPane.rightAnchor="40.0" />
         <Label alignment="CENTER" text="Welcome" textAlignment="CENTER" AnchorPane.leftAnchor="100.0" AnchorPane.rightAnchor="100.0" AnchorPane.topAnchor="20.0">
             <font>
                 <Font size="32.0" />
             </font>
         </Label>
-      <Label alignment="CENTER" text="Please choose a database to open, or press Control+O to open a new file" textAlignment="CENTER" AnchorPane.leftAnchor="100.0" AnchorPane.rightAnchor="100.0" AnchorPane.topAnchor="80.0" />
+      <Label alignment="CENTER" text="Please choose a database to open, or press Ctrl+O to open a new file" textAlignment="CENTER" AnchorPane.leftAnchor="100.0" AnchorPane.rightAnchor="100.0" AnchorPane.topAnchor="80.0" />
```
Comment (Balletie) [inline dnainator-javafx/src/main/resources/fxml/welcome.fxml]: Should this still be rephrased to say "choose a directory with databases", or something along those lines?

**Comment by Balletie** [comment]
No further comments, so merging.
